### PR TITLE
Prometheus bugfixes

### PR DESCRIPTION
**Pull request by kryztoval**
Fixed an unintended exception/logic error when a Type was not registered in the dictionary
Added distinctive aliases for hardware and sensors to prevent name collision due to duplicated names under hardware or sensors as mentioned in #1925
Reviewed units and factors
Skip null values and show an error message as per #1979

### CODE REVIEW

File: LibreHardwareMonitor/Utilities/HttpServer.cs [hunk 1]
```diff
@@ -610,38 +610,87 @@ private string GeneratePrometheusResponse(Node node)
...
                 {
-                    string tagHardware = (((HardwareNode)node).Hardware.Parent != null ? ((HardwareNode)node).Hardware.Parent.HardwareType.ToString() : ((HardwareNode)node).Hardware.HardwareType.ToString());
+                    // Variables needed in dictionary lookup and error message
                     string tagSensorType = sensor.Sensor.SensorType.ToString();
-                    string tagSensorUnits = (units[sensor.Sensor.SensorType].Item1.Length == 0 ? "" : "_" + units[sensor.Sensor.SensorType].Item1);
+                    string valueSensor = sensor.Text.Replace("#", String.Empty);
+                    string tagHardware = (((HardwareNode)node).Hardware.Parent != null ? ((HardwareNode)node).Hardware.Parent.HardwareType.ToString() : ((HardwareNode)node).Hardware.HardwareType.ToString());
+
+                    double _factor = 1;
+                    string tagSensorUnits = String.Empty;
```
Comment (PhyxionNL) [inline LibreHardwareMonitor/Utilities/HttpServer.cs]:
```suggestion
                    string tagSensorUnits = "";
```

File: LibreHardwareMonitor/Utilities/HttpServer.cs [hunk 2]
```diff
@@ -610,38 +610,87 @@ private string GeneratePrometheusResponse(Node node)
...
+                    // Get factor and unit suffix from dictionary ...
+                    if (units.ContainsKey(sensor.Sensor.SensorType))
+                    {
+                        _factor = units[sensor.Sensor.SensorType].Item2;
+                        tagSensorUnits = (units[sensor.Sensor.SensorType].Item1.Length == 0 ? String.Empty : "_" + units[sensor.Sensor.SensorType].Item1);
+                    }
+                    // ... or print an error message
+                    else { responseStr += $"# HELP {tagHardware}_{tagSensorType}:{valueSensor} This Sensor type is not defined in the prometheus adapter [{sensor.Sensor.SensorType}]\n"; }
```
Comment (PhyxionNL) [inline LibreHardwareMonitor/Utilities/HttpServer.cs]: Put this in 3 lines please.

File: LibreHardwareMonitor/Utilities/HttpServer.cs [hunk 3]
```diff
@@ -610,38 +610,87 @@ private string GeneratePrometheusResponse(Node node)
...
+
+                    // Preparing the labels for all data and uniqueness
+                    string valueSensorId = sensor.Sensor.Identifier.ToString();
                     string valueHardware = node.Text;
-                    string valueId = sensor.Sensor.Identifier.ToString().Split('/').Last();
-                    string valueFamily = node.Nodes[i].Text;
-                    string valueHost = _root.Text;
 
-                    try
+                    string valueHardwareAlias = String.Empty;
```
Comment (PhyxionNL) [inline LibreHardwareMonitor/Utilities/HttpServer.cs]: This seems far too complicated to just guarantee unique names. Perhaps just use the existing logic and add all ids to a Dictionary<string, int>? If it doesn't exist, use it as-is. If it exists, increment and add this int to the identifier.

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: The idea is not to guarantee that the names are unique, the point is to guarantee the names do not repeat when they should not.

A GPU can have 2 "Video Decoders". Storage can have 2 of the same "XPG GAMMIX S41" drives.

Hardware can be nested. I would need to have a unique identifier for the Hardware to make sure the Name is not repeated. But this makes the same problem I am having right now.

I will need to think about this a bit more. Give me a few hours to think this thru.

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: I am now using dictionaries to keep track of the names. 
`hardware` and `hardware.sensor` should be unique. 

I look up the sensor ID in the sensor <str ID,str name> dictionary. If I found it i can just use the name.
If it has not been created yet I look at the names <str id, int next_number>  for a `hardware.sensor`, either adding it or adding the number and incrementing it by one.
Then I save this new name to the  sensor<str ID, str name> dictionary.

Seems still quite complicated, but as I said. The point was not just making sure names were unique, the point was making sure that the name is only changed when it is duplicated and at the level where it was duplicated.

I moved this to be a draft to test this code in more computers with duplicated hardware and for comments on the new code.

**Comment by kryztoval** [comment]
Even though I am now using dictionaries for this I feel that this is still more complicated than it seems. Just adding  an incremental number to the name doesn't work.

It is now clearer, has comments, uses Dictionaries, solves the 2 edge cases (2 hardware with the same name. 2 sensors with the same name).

File: LibreHardwareMonitor/Utilities/HttpServer.cs (hunk 3 above)
Comment (PhyxionNL) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: Hardware should not be an issue as they _should_ have unique Identifier (Storage/GPU has this for example). Their sensor names can contain identical names, but a simple local dictionary with a count should be OK as they're always added in the same order.

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: Yes, that was the theory, but in practice this is not the case. 

For an example I do have 2x XPG GAMMIX S41 and both appear with the exact same name.

So hardware is an issue and that is why I have to be careful with that. I noticed because I was using the hardware name as a grouping for the tags in grafana and it turned out that I had duplicated tags because I had 2 hardware devices that are identical.

Sensor names can contain identical names in the same hardware. And this case would be solved with a simple dictionary. 

However the duplicated hardware name is a problem.

Here you can see that nvme/1 and nvme/2 have the same name and were replaced by the code I submitted.

This seems like an edge case, not many people would have identical hardware. but I can probably test it with 2x 3090s too. To see if the hardware name changes.

Also of importance is that the hardware name change remains consistent for all sensors of that device, this is why I made 2 dictionaries, one for the paths and one for the names.

```
# TYPE lhm_storage_temperature_celsius gauge
lhm_storage_temperature_celsius {"sensor"="Temperature" "hardware"="XPG GAMMIX S41" "id"="/nvme/1/temperature/0" "parentId"="/nvme/1" "family"="Temperatures" "host"="[REDACTED]"} 40
# TYPE lhm_storage_load_percent gauge
lhm_storage_load_percent {"sensor"="Used Space" "hardware"="XPG GAMMIX S41" "id"="/nvme/1/load/0" "parentId"="/nvme/1" "family"="Load" "host"="[REDACTED]"} 16.5353546142578
lhm_storage_load_percent {"sensor"="Read Activity" "hardware"="XPG GAMMIX S41" "id"="/nvme/1/load/31" "parentId"="/nvme/1" "family"="Load" "host"="[REDACTED]"} 0
lhm_storage_load_percent {"sensor"="Write Activity" "hardware"="XPG GAMMIX S41" "id"="/nvme/1/load/32" "parentId"="/nvme/1" "family"="Load" "host"="[REDACTED]"} 1.71433210372925
lhm_storage_load_percent {"sensor"="Total Activity" "hardware"="XPG GAMMIX S41" "id"="/nvme/1/load/33" "parentId"="/nvme/1" "family"="Load" "host"="[REDACTED]"} 1.10831892490387
# TYPE lhm_storage_level_percent gauge
lhm_storage_level_percent {"sensor"="Available Spare" "hardware"="XPG GAMMIX S41" "id"="/nvme/1/level/1" "parentId"="/nvme/1" "family"="Levels" "host"="[REDACTED]"} 94
lhm_storage_level_percent {"sensor"="Available Spare Threshold" "hardware"="XPG GAMMIX S41" "id"="/nvme/1/level/2" "parentId"="/nvme/1" "family"="Levels" "host"="[REDACTED]"} 32
lhm_storage_level_percent {"sensor"="Percentage Used" "hardware"="XPG GAMMIX S41" "id"="/nvme/1/level/3" "parentId"="/nvme/1" "family"="Levels" "host"="[REDACTED]"} 0
# TYPE lhm_storage_data_bytes gauge
lhm_storage_data_bytes {"sensor"="Data Read" "hardware"="XPG GAMMIX S41" "id"="/nvme/1/data/4" "parentId"="/nvme/1" "family"="Data" "host"="[REDACTED]"} 34377000000000
lhm_storage_data_bytes {"sensor"="Data Written" "hardware"="XPG GAMMIX S41" "id"="/nvme/1/data/5" "parentId"="/nvme/1" "family"="Data" "host"="[REDACTED]"} 3815000000000
# TYPE lhm_storage_throughput_bytes_per_second gauge
lhm_storage_throughput_bytes_per_second {"sensor"="Read Rate" "hardware"="XPG GAMMIX S41" "id"="/nvme/1/throughput/34" "parentId"="/nvme/1" "family"="Throughput" "host"="[REDACTED]"} 0
lhm_storage_throughput_bytes_per_second {"sensor"="Write Rate" "hardware"="XPG GAMMIX S41" "id"="/nvme/1/throughput/35" "parentId"="/nvme/1" "family"="Throughput" "host"="[REDACTED]"} 275009.09375
# TYPE lhm_storage_temperature_celsius gauge
lhm_storage_temperature_celsius {"sensor"="Temperature" "hardware"="XPG GAMMIX S41 1" "id"="/nvme/2/temperature/0" "parentId"="/nvme/2" "family"="Temperatures" "host"="[REDACTED]"} 37
# TYPE lhm_storage_load_percent gauge
lhm_storage_load_percent {"sensor"="Used Space" "hardware"="XPG GAMMIX S41 1" "id"="/nvme/2/load/0" "parentId"="/nvme/2" "family"="Load" "host"="[REDACTED]"} 16.5353546142578
lhm_storage_load_percent {"sensor"="Read Activity" "hardware"="XPG GAMMIX S41 1" "id"="/nvme/2/load/31" "parentId"="/nvme/2" "family"="Load" "host"="[REDACTED]"} 0
lhm_storage_load_percent {"sensor"="Write Activity" "hardware"="XPG GAMMIX S41 1" "id"="/nvme/2/load/32" "parentId"="/nvme/2" "family"="Load" "host"="[REDACTED]"} 1.07297778129578
lhm_storage_load_percent {"sensor"="Total Activity" "hardware"="XPG GAMMIX S41 1" "id"="/nvme/2/load/33" "parentId"="/nvme/2" "family"="Load" "host"="[REDACTED]"} 1.05270564556122
# TYPE lhm_storage_level_percent gauge
lhm_storage_level_percent {"sensor"="Available Spare" "hardware"="XPG GAMMIX S41 1" "id"="/nvme/2/level/1" "parentId"="/nvme/2" "family"="Levels" "host"="[REDACTED]"} 100
lhm_storage_level_percent {"sensor"="Available Spare Threshold" "hardware"="XPG GAMMIX S41 1" "id"="/nvme/2/level/2" "parentId"="/nvme/2" "family"="Levels" "host"="[REDACTED]"} 32
lhm_storage_level_percent {"sensor"="Percentage Used" "hardware"="XPG GAMMIX S41 1" "id"="/nvme/2/level/3" "parentId"="/nvme/2" "family"="Levels" "host"="[REDACTED]"} 4
# TYPE lhm_storage_data_bytes gauge
lhm_storage_data_bytes {"sensor"="Data Read" "hardware"="XPG GAMMIX S41 1" "id"="/nvme/2/data/4" "parentId"="/nvme/2" "family"="Data" "host"="[REDACTED]"} 2972000000000
lhm_storage_data_bytes {"sensor"="Data Written" "hardware"="XPG GAMMIX S41 1" "id"="/nvme/2/data/5" "parentId"="/nvme/2" "family"="Data" "host"="[REDACTED]"} 80720000000000
# TYPE lhm_storage_throughput_bytes_per_second gauge
lhm_storage_throughput_bytes_per_second {"sensor"="Read Rate" "hardware"="XPG GAMMIX S41 1" "id"="/nvme/2/throughput/34" "parentId"="/nvme/2" "family"="Throughput" "host"="[REDACTED]"} 0
lhm_storage_throughput_bytes_per_second {"sensor"="Write Rate" "hardware"="XPG GAMMIX S41 1" "id"="/nvme/2/throughput/35" "parentId"="/nvme/2" "family"="Throughput" "host"="[REDACTED]"} 275069.46875
```

Comment (PhyxionNL) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: > For an example I do have 2x XPG GAMMIX S41 and both appear with the exact same name.

But their Identifiers _will_ be different, so just group by name + id? I don't think it makes a lot of sense to change device names to 1 and 2 for example. With sensors, this is different as with D3D there are multiple "engines" with the same name. Windows also lists these as Copy, Copy 1, Copy 2, but with two storage devices, they're not named Device Name 1 and Device Name 2.

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: Yes, I agree, it makes almost no sense to LHM or in the context of C#/Windows. The context is provided by the intended client of this data: prometheus and grafana.

The list of problems that this solves are all grafana's restrictions.

In grafana:

- The `dropdown`'s `value` and `text` are the same. So using the `hardware.id` alone would provide no clues to what the user is selecting. This also makes it really cumbersome to have all elements without duplicates to have an extra number at the end.
- You can only make a list from only one label. This prevents using `hardware.name + hardware.id` and forces `hardware.name` to be unique.
- The repeating rows can only filter on one label. This enforces `hardware.name` uniqueness again. 

With this change a dashboard that lists all the sensors per hardware and adds/remove rows per hardware becomes automatic.
Also, with the sensor renaming it shows different lines in the graph instead of incorrectly merging them together.

This change is only required because of grafana's limitations on filtering the data from prometheus. And it would be extremely useful for the users of LHM that have a prometheus data backend.

What can I change to this so we can have unique hardware names in the prometheus export?

Comment (PhyxionNL) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: That sounds like a recipe for disaster if you replace hardware with an identical type and put it in another slot.
What about something like this: 
- hardware.description > device.name.
- hardware.name > device.name (device.id).
- hardware.id > device.id

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: I do not understand this message. 
What recipe? what disaster?

Anyways, in prometheus we have a tag name which can be considered the generic tag that stores the data. for example: `lhm_cpu_voltage_volts`. Which is composed of:
- lhm: The source
- cpu: The hardware type
- voltage: What is being measured
- volts: The units

This is a dimension that we can store values to. the values in prometheus are unique based on labels. so `lhm_cpu_voltage_volts { sensor=/cpu/0 hardware=Intel }` is a different data trend than `lhm_cpu_voltage_volts { sensor=/cpu/1 harddware=Intel }`

Originally I was exporting the following labels as-is `sensor, hardware, id, family, host`:
- sensor: the sensor name, used as the description of the plot
- hardware: the hardware name, used as the device that hosts the sensors
- id: the sensor count, which resets on each hardware
- host: the computer name that hosts the hardware
- family: the sensor type, not really used

What I tried in this commit was to add a hardwareAlias and a sensorAlias that are a rename with uniqueness of the original as-is hardware and sensor names. You said the logic behind this was too complicated and that it could use a Dictionary. I agree. I was doing some path calculations and merging splitting paths in a cumbersome way.

The current proposition was to replace the hardware and sensor names with their aliases. The changes would be:
- sensor: (replaced by sensorAlias)  sensor name plus a number if it is repeated in the same hardware group
- hardware: (replaced by hardwareAlias) hardware name plus a number if it is repeated in the same host group
- id: the sensor path, for tracking changes, not used just easier to debug. The original value was not useful since it is not unique globally, only locally.
- parentId: the hardware path, for tracking changes, not used or required.
- all others remain unchanged

This changes in name are only done in the export of the text. The dictionaries are only kept at runtime because prometheus doesn't care if the name changes.

I do not understand this part
> * hardware.description > device.name.
> * hardware.name > device.name (device.id).
> * hardware.id > device.id

Could you elaborate?

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: > Windows also lists these as Copy, Copy 1, Copy 2, but with two storage devices, they're not named Device Name 1 and Device Name 2.

Windows names "D3D Copy"  engines as "D3D Copy" and "D3D Copy 1". But windows names the drives as Disk0 and Disk1. Because the engines are already the description then windows renames them when there are duplicates. But because hardware is a description in windows it is not renamed nor does it need to be. You can always reference the device uniquely by its id or name. This is not the case in prometheus where the name is meant to be user facing and unique.

Comment (PhyxionNL) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: > Could you elaborate?

I mean adding another data field with a description ([Hardware.Name](https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/blob/master/LibreHardwareMonitorLib/Hardware/Hardware.cs#L52)) and changing the hardware name to [Hardware.Name](https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/blob/master/LibreHardwareMonitorLib/Hardware/Hardware.cs#L52) ([Hardware.Identifier](https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/blob/master/LibreHardwareMonitorLib/Hardware/Hardware.cs#L49)), so then you'll always have a unique bit.

So in your example something like:
```txt
"hardware"="XPG GAMMIX S41 (/nvme/2)"
"hardwareDesc"="XPG GAMMIX S41" 
"id"="/nvme/2/throughput/34" 
"parentId"="/nvme/2"
``` 

If you're going to count things yourself, this won't be reliable if you move hardware around.

> But windows names the drives as Disk0 and Disk1.

Those are not names but disk ids (see diskpart).

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: > I mean adding another data field with a description ([Hardware.Name](https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/blob/master/LibreHardwareMonitorLib/Hardware/Hardware.cs#L52)) and changing the hardware name to [Hardware.Name](https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/blob/master/LibreHardwareMonitorLib/Hardware/Hardware.cs#L52) ([Hardware.Identifier](https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/blob/master/LibreHardwareMonitorLib/Hardware/Hardware.cs#L49)), so then you'll always have a unique bit.

And in this case, do you propose this to apply to every device? even the ones that have no duplicates?

While that would work it would add extra noise to every single hardware list in prometheus, not just the ones that have duplicates. It would also add an extra label that will be a copy of two already existing fields (`hardware` and `parentId`). Prometheus can merge those fields to show them if it wanted to, but it can not remove parts of a field in a dropdown. 

> If you're going to count things yourself, this won't be reliable if you move hardware around.

The "/nvme/2" also changes if the device is moved around. The id on the data.json endpoint is an incremental int. Both of those options are equally as reliable and I would argue that it counts things themselves.

The objective of counting things myself is to guarantee that the name is not duplicated while not changing anything else if there are no duplicates.  And I did export the `sensor.Identifier` as the label `id` to guarantee at least the same level of stability.

Prometheus by default stores at most 2 weeks of data, and the correlation can be mapped to another label. For example, if you had a sensorid `/nvme/2` stored in the prometheus label `id` even if the prometheus label `hardware` change its value you could group them by the prometheus label `sensorid`.

Prometheus also fills the drop downs and options from the time range. This implies the time window has to be contained between 5 seconds and 2 weeks.

> Those are not names but disk ids (see diskpart).

That *is* the name, and by name I mean an identifier that is unique and can be used to select the device. If you use `list disk` in `diskpart` you will see there is no "name" field. the only name available is "Disk ###". All other fields are either a value or a status. Then you can select it by issuing `select disk 1`. But you cannot select it by its description "XPG GAMMIX S41" nor by any of the other values returned.

However, if you list the sensors in windows you can select them by "name" or by "id".



How about this? I pad the `name` label with 2 spaces at the end. and if it has duplicates I write the number there. So you can always get the original name by removing 2 spaces from the name in the export in a reliable and predictable way that will not add extra labels, reduces the amount of space taken on export and is predictable.

File: LibreHardwareMonitor/Utilities/HttpServer.cs [hunk 4]
```diff
@@ -672,46 +676,46 @@ private string GeneratePrometheusResponse(Node node)
...
-                        _prometheusPaths.Add(valueId, $"{valueSensor}");
+                        _prometheusIdName.Add(valueSensorId, $"{valueSensorName}");
                     }
 
                     string valueFamily = node.Nodes[i].Text;
                     string valueHost = _root.Text;
 
                     // Creates the tag with labels
-                    string tagLine = $$"""{{tagName}} {"sensor"="{{valueSensor}}" "hardware"="{{valueHardware}}" "id"="{{valueId}}" "parentId"="{{hardwarePath}}" "family"="{{valueFamily}}" "host"="{{valueHost}}"}""";
+                    string tagLine = $$"""{{tagName}} {"sensorName"="{{valueSensorName}}" "sensorAlias"="{{valueSensorAlias}}" "hardwareName"="{{valueHardwareName}}" "hardwareAlias"="{{valueHardwareAlias}}" "sensorId"="{{valueSensorId}}" "hardwareId"="{{valueHardwareId}}" "host"="{{valueHost}}"}""";
```
Comment (kryztoval) [inline LibreHardwareMonitor/Utilities/HttpServer.cs]: I do not like the idea of having `hardwareAlias` be `{valueHardwareName} ({valueHardwareId})` and `sensorAlias` be `{valueSensorName} ({valueSensorId})`

Pros: It requires no dictionaries, all the naming logic can be removed.
Cons: it modifies all the aliases not just the duplicates, can't be modified downstream, some ids are really long (e.g. 
/throughput/7/nic/%7B2DCDD9D7-D331-466D-926F-D6C957B2D917%7D)

But if that is the only way for this PR to go through then I will do that because I would rather have the data in Prometheus rather than not having it.

File: LibreHardwareMonitor/Utilities/HttpServer.cs (hunk 3 above)
Comment (PhyxionNL) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: > How about this? I pad the name label with 2 spaces at the end. and if it has duplicates I write the number there. So you can always get the original name by removing 2 spaces from the name in the export in a reliable and predictable way that will not add extra labels, reduces the amount of space taken on export and is predictable.

I don't see how this is better than just adding ```Identifier``` to the name. If you want to get the original name you can simplify replace everything between brackets. We don't have to add ```hardwareDesc``` if there's no use case.

With various hardware this will be more reliable than manually counting. For example DIMM uses the physical slot number as identifier.

So basically:
```` 
"hardware"="XPG GAMMIX S41 (/nvme/2)"
````

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: This is how it is better than just adding `Identifier` to the name.
- An empty space is automatically removed by html rendering and causes no visual clutter in Grafana while having " (/sensor/path)" will add visual clutter to all of the elements
- All elements that are not duplicates will have a consistent name in Prometheus and the data.json endpoint and the LHM interface
- Only items with duplicated names would have a visual differentiator added to them
- The labels are attached to the value, you can always find the hardwareId and sensorId on a data point. You can form the plot description combining more than one label (HardwareName and HardwareId)
- Avoids Grafana's restriction that prevents us from removing text from a label or merging more than one label inside a selector


> If you want to get the original name you can simplify replace everything between brackets.

As simple as this sounds, this can't be done in Prometheus/Grafana. You can extract parts of the tagName to return a list of tag names, or return all distinct values of a single label from a group of tags with the name name being regex and/or the labels being filtered by an exact match.

> With various hardware this will be more reliable than manually counting. For example DIMM uses the physical slot number as identifier.

DIMM actually adds `#0`, `#1`, `#2`, `#4`, etc. To the memory name if they have the same name. Manual counting does not happen on those devices because their names are already different.

Manual counting seems like the best option because it currently only happens on Nodes that are of type Hardware or Sensor that have siblings with the same name. Nowhere else.

We still have unreliable counting elsewhere, when you move hardware around the `(/sensor/path)` is not preserved either. And the `id` number that is shown in the `/data.json` endpoint is an incremental line counter integer. So adding a single device will increment all lines coming after it by the number of sensors (... or lines) that the device added.

The only cons I see are that it requires extra logic and 2 dictionaries to guarantee the names to be unique. Prometheus doesn't care if the names are consistent across reboots. The dictionaries do not need to be stored across restarts either. Grafana can complain about the names being too long specially given how some devices ids are really long (e.g. nic devices)

File: LibreHardwareMonitor/Utilities/HttpServer.cs (hunk 4 above)
Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: Ok, made it so it is `name (id)`

This should be ready to merge.

**Comment by kryztoval** [comment]
Just to be extremely clear, I hate the `name (id)` thing.

image

File: LibreHardwareMonitor/Utilities/HttpServer.cs [hunk 5]
```diff
@@ -610,37 +612,77 @@ private string GeneratePrometheusResponse(Node node)
...
                     {
+                        responseStr += $"# TYPE {tagName} gauge\n";
+                        lastTagName = tagName;
+                    }
 
-                        if (lastTagName != tagName)
+                    int _counter = 0;
+                    foreach (SensorValue val in sensor.Sensor.Values.Reverse())
+                    {
+                        if (_counter++ >= _maxValues) break;
```
Comment (PhyxionNL) [inline LibreHardwareMonitor/Utilities/HttpServer.cs]: Why do you need a history of sensor values?

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: Prometheus is a time series database, by default it stores the last 14 days of values by scraping the `/metrics` endpoint on a user defined interval.

In the original pull request it only showed the last value without a timestamp and it would be stored at the time it was read, not the actual measurement time. In this pull request the values are sent with their respective timestamps.

This means you have more leniency and can scrape the endpoint less often. The theory on rate scraping says you have to sample something at twice the speed to be sure you have all the original values (e.g. Music can be heard by humans at most at 22050Hz, so CD Audio format would sample it at 44100Hz). 

Having 5 previous values with their timestamps means you can scrape 4x slower than the sampling rate and still have a 100% confidence that you collected all the values.

It was considered since the first prometheus endpoint pull request as seen in the notes of "known issues", but I didn't know at the time that there was a history of values per sensor with timestamps that I could use.

I limited this to 5 and only sent the most recent values.

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: Here you can see the trends generated by Grafana from Prometheus as exported by this pull request. A few blanks are in the graph because the computer was restarted and there was no data for the time.

You can clearly see the drop downs at the top with the alias. You can see why I didn't want to use the id as part of the alias.

image

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: The answer is still valid @PhyxionNL.

If I do not have new feedback on a question that I already answered then I do not know how to better answer it.

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: I will rewrite it here because you might have missed it

> Having 5 previous values with their timestamps means you can scrape 4x slower than the sampling rate and still have a 100% confidence that you collected all the values.

If sensor updates every second, and the /metrics endpoint returns 5 seconds, you can scrape it every 4 seconds and still get all values. Returning the historic values is cheaper because all the logic that has to be performed for each scrape is skipped for the extra 4 values and only the multiplication is done.

Why would we not want a history of the values?

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: Benefits of having a history of N sensor values of a given Interval
- Allows you to poll at `(N-1) x Interval` and still get all the values
- Polling once every `(N-1) x Interval` is less expensive than polling once every `Interval`
- Reduces used bandwidth
- Makes it possible to recover missed values
- Uses an already existing history of sensor values
- Better represents the data acquisition timestamps
- Doesn't add complexity to the code

From my perspective I see no reason not to send 5 values as a response.

Why not send 5 values as a response if everything is there already?

Comment (PhyxionNL) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: @kryztoval ^^

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: ^^

Comment (PhyxionNL) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: This is not resolved @kryztoval.

Comment (kryztoval) [reply LibreHardwareMonitor/Utilities/HttpServer.cs]: I wrote 5 messages with a detailed and extensive explanation. And all I am getting from you is a "this is not solved'" but not a single explanation on why you think that the reasoning behind this is invalid. 

I also allowed 2 weeks for you to read and respond to those comments, your response was ”^^". I am sorry, but from my point of view this is more than resolved. You are giving me nothing to work with. 

Because it feels like you have either not read, refused to acknowledge or have not reviewed the responses in this thread I am forced to re-reauedt a review instead of answering something I already answered at length.

**Comment by kryztoval** [comment]
The response is still valid.

File: LibreHardwareMonitor/Utilities/HttpServer.cs [hunk 6]
```diff
@@ -571,11 +571,13 @@ private async Task SendJsonAsync(HttpListenerResponse response, HttpListenerRequ
 
     private string GeneratePrometheusResponse(Node node)
     {
+        const int _maxValues = 5;
```
Comment (PhyxionNL) [inline LibreHardwareMonitor/Utilities/HttpServer.cs]:
```suggestion
        const int MaxValues = 5;
```

File: LibreHardwareMonitor/Utilities/HttpServer.cs [hunk 7]
```diff
@@ -610,37 +612,77 @@ private string GeneratePrometheusResponse(Node node)
 
             if (node.Nodes[i].GetType().Name == "TypeNode")
             {
-                string prometheusHost = node.Parent.Text;
+
```
Comment (PhyxionNL) [inline LibreHardwareMonitor/Utilities/HttpServer.cs]:
```suggestion
```

File: LibreHardwareMonitor/Utilities/HttpServer.cs [hunk 8]
```diff
@@ -610,37 +612,77 @@ private string GeneratePrometheusResponse(Node node)
...
+                string tagHardware = "";
+                string valueHardwareName = "";
+                string valueHardwareId = ((HardwareNode)node).Hardware.Identifier.ToString();
+                if (((HardwareNode)node).Hardware.Parent != null)
```
Comment (PhyxionNL) [inline LibreHardwareMonitor/Utilities/HttpServer.cs:619]:
```suggestion

                if (((HardwareNode)node).Hardware.Parent != null)
```

**Comment by kryztoval** [comment]
I rebased it, not sure if I did it correctly.

**Comment by kryztoval** [comment]
I think this means all of these issues can be closed now #1138 #1293  #1624 #1648 #1997

**Comment by kryztoval** [comment]
@PhyxionNL pinging you just in case you missed the message above. 🎄 Happy holidays!

**Comment by PhyxionNL** [comment]
Thanks, you too @kryztoval!
